File: lib/projection/projection.c
#include "yai/projection/projection.h"

#include <stdio.h>
#include <string.h>

static int summary_has_memory_kind(const yai_store_record_t *record,
                                   const char *kind) {
    char pattern[64];

    if (record == 0 || kind == 0) {
        return 0;
    }
    (void)snprintf(pattern, sizeof(pattern), "memory:%s", kind);
    return strstr(record->summary, pattern) != 0;
}

yai_status_t yai_projection_build(const char *projection_id,
                                  const yai_case_ref_t *case_ref,
                                  yai_projection_consumer_t consumer_kind,
                                  const yai_journal_t *journal,
                                  yai_projection_t *projection) {
    size_t index = 0;

    if (projection == 0 || projection_id == 0 || projection_id[0] == '\0' ||
        yai_case_ref_is_empty(case_ref) || journal == 0) {
        return YAI_ERR_INVALID;
    }

    yai_id_set(&projection->projection_id, projection_id);
    projection->case_ref = *case_ref;
    projection->consumer_kind = consumer_kind;
    projection->source_record_count = yai_journal_count(journal);
    projection->decision_count = 0;
    projection->policy_rule_count = 0;
    projection->gate_count = 0;
    projection->obligation_count = 0;
    projection->receipt_requirement_count = 0;
    projection->filesystem_receipt_count = 0;
    projection->subject_state_count = 0;
    projection->effect_count = 0;
    projection->graph_edge_count = 0;
    projection->reconstruction_count = 0;
    projection->memory_candidate_count = 0;
    projection->operational_memory_candidate_count = 0;
    projection->decision_memory_candidate_count = 0;
    projection->subject_memory_candidate_count = 0;
    projection->error_memory_candidate_count = 0;
    projection->recovery_memory_candidate_count = 0;
    projection->divergence_count = 0;
    projection->reconciliation_count = 0;
    projection->critical_divergence_count = 0;
    projection->warning_divergence_count = 0;

    for (index = 0; index < yai_journal_count(journal); index += 1) {
        const yai_store_record_t *record = yai_journal_get(journal, index);
        if (record == 0) {
            continue;
        }
        if (record->record_kind == YAI_RECORD_DECISION) {
            projection->decision_count += 1;
        } else if (record->record_kind == YAI_RECORD_POLICY_RULE) {
            projection->policy_rule_count += 1;
        } else if (record->record_kind == YAI_RECORD_GATE_RESULT) {
            projection->gate_count += 1;
        } else if (record->record_kind == YAI_RECORD_OBLIGATION) {
            projection->obligation_count += 1;
        } else if (record->record_kind == YAI_RECORD_RECEIPT_REQUIREMENT) {
            projection->receipt_requirement_count += 1;
        } else if (record->record_kind == YAI_RECORD_FILESYSTEM_RECEIPT) {
            projection->filesystem_receipt_count += 1;
            projection->effect_count += 1;
        } else if (record->record_kind == YAI_RECORD_EFFECT_RECEIPT) {
            projection->effect_count += 1;
        } else if (record->record_kind == YAI_RECORD_SUBJECT_STATE) {
            projection->subject_state_count += 1;
        } else if (record->record_kind == YAI_RECORD_RECEIPT) {
            projection->effect_count += 1;
        } else if (record->record_kind == YAI_RECORD_GRAPH_EDGE) {
            projection->graph_edge_count += 1;
        } else if (record->record_kind == YAI_RECORD_RECONSTRUCTION) {
            projection->reconstruction_count += 1;
        } else if (record->record_kind == YAI_RECORD_MEMORY_CANDIDATE) {
            projection->memory_candidate_count += 1;
            if (summary_has_memory_kind(record, "operational")) {
                projection->operational_memory_candidate_count += 1;
            } else if (summary_has_memory_kind(record, "decision")) {
                projection->decision_memory_candidate_count += 1;
            } else if (summary_has_memory_kind(record, "subject")) {
                projection->subject_memory_candidate_count += 1;
            } else if (summary_has_memory_kind(record, "error")) {
                projection->error_memory_candidate_count += 1;
            } else if (summary_has_memory_kind(record, "recovery")) {
                projection->recovery_memory_candidate_count += 1;
            }
        } else if (record->record_kind == YAI_RECORD_DIVERGENCE) {
            projection->divergence_count += 1;
            if (strstr(record->summary, "severity:critical") != 0) {
                projection->critical_divergence_count += 1;
            } else if (strstr(record->summary, "severity:warning") != 0) {
                projection->warning_divergence_count += 1;
            }
        } else if (record->record_kind == YAI_RECORD_RECONCILIATION) {
            projection->reconciliation_count += 1;
        }
    }

    (void)snprintf(projection->summary,
                   sizeof(projection->summary),
                   "projection:%s records:%zu decisions:%zu rules:%zu gates:%zu obligations:%zu receipt_requirements:%zu filesystem_receipts:%zu subject_states:%zu effects:%zu graph_edges:%zu reconstructions:%zu memory_candidates:%zu divergences:%zu reconciliations:%zu",
                   yai_projection_consumer_string(consumer_kind),
                   projection->source_record_count,
                   projection->decision_count,
                   projection->policy_rule_count,
                   projection->gate_count,
                   projection->obligation_count,
                   projection->receipt_requirement_count,
                   projection->filesystem_receipt_count,
                   projection->subject_state_count,
                   projection->effect_count,
                   projection->graph_edge_count,
                   projection->reconstruction_count,
                   projection->memory_candidate_count,
                   projection->divergence_count,
                   projection->reconciliation_count);
    return YAI_OK;
}

const char *yai_projection_consumer_string(yai_projection_consumer_t consumer) {
    switch (consumer) {
    case YAI_PROJECTION_OPERATOR:
        return "operator";
    case YAI_PROJECTION_MODEL:
        return "model";
    case YAI_PROJECTION_AGENT:
        return "agent";
    case YAI_PROJECTION_API:
        return "api";
    case YAI_PROJECTION_AUDIT:
        return "audit";
    case YAI_PROJECTION_DEBUG:
        return "debug";
    default:
        return "unknown";
    }
}

File: crates/yai-core-engine/src/projection.rs
use crate::journal::Journal;
use crate::record::RecordKind;
use std::collections::BTreeSet;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ProjectionSummary {
    pub consumer: String,
    pub case_ref: String,
    pub summary: String,
    pub source_record_count: usize,
    pub receipt_count: usize,
    pub subject_count: usize,
    pub decision_count: usize,
    pub policy_rule_count: usize,
    pub gate_count: usize,
    pub obligation_count: usize,
    pub receipt_requirement_count: usize,
    pub filesystem_receipt_count: usize,
    pub subject_state_count: usize,
    pub effect_count: usize,
}

impl ProjectionSummary {
    pub fn from_journal(consumer: impl Into<String>, journal: &Journal) -> Self {
        let consumer = consumer.into();
        let source_record_count = journal.count();
        let case_ref = journal
            .records()
            .iter()
            .find(|record| !record.case_ref.is_empty())
            .map(|record| record.case_ref.clone())
            .unwrap_or_default();
        let receipt_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::Receipt)
            .count();
        let decision_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::Decision)
            .count();
        let policy_rule_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::PolicyRule)
            .count();
        let gate_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::GateResult)
            .count();
        let obligation_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::Obligation)
            .count();
        let receipt_requirement_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::ReceiptRequirement)
            .count();
        let filesystem_receipt_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::FilesystemReceipt)
            .count();
        let subject_state_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::SubjectState)
            .count();
        let effect_count = journal
            .records()
            .iter()
            .filter(|record| {
                matches!(
                    record.kind,
                    RecordKind::Receipt | RecordKind::EffectReceipt | RecordKind::FilesystemReceipt
                )
            })
            .count();
        let subject_count = journal
            .records()
            .iter()
            .filter(|record| !record.subject_ref.is_empty() && record.subject_ref != "subject:none")
            .map(|record| record.subject_ref.as_str())
            .collect::<BTreeSet<_>>()
            .len();
        Self {
            summary: format!(
                "projection:{consumer} records:{source_record_count} decisions:{decision_count} rules:{policy_rule_count} gates:{gate_count} obligations:{obligation_count} receipt_requirements:{receipt_requirement_count} filesystem_receipts:{filesystem_receipt_count} subject_states:{subject_state_count} effects:{effect_count}"
            ),
            consumer,
            case_ref,
            source_record_count,
            receipt_count,
            subject_count,
            decision_count,
            policy_rule_count,
            gate_count,
            obligation_count,
            receipt_requirement_count,
            filesystem_receipt_count,
            subject_state_count,
            effect_count,
        }
    }
}
